plugins/out_doris/doris.c

```diff
@@ -0,0 +1,461 @@
+                        break;
+                    }
+                    
+                    out_ret = FLB_RETRY;
+                    break;
+                }
+            }
+        
+            flb_free(out_buf);
+            msgpack_unpacked_destroy(&result);
+        }
+        else {
+            out_ret = FLB_RETRY;
+        }
+    }
+    else {
+        flb_plg_error(ctx->ins, "could not flush records to %s:%i (http_do=%i)",
+                      ctx->host, ctx->port, ret);
+        out_ret = FLB_RETRY;
+    }
+
+    /* cleanup */
+    
+    /*
+     * If the payload buffer is different than incoming records in body, means
+     * we generated a different payload and must be freed.
+     */
+    if (payload_buf != body) {
+        flb_free(payload_buf);
+    }
+
+    /* Destroy HTTP client context */
+    flb_http_client_destroy(c);
+
+    /* Release the TCP connection */
+    flb_upstream_conn_release(u_conn);
+
+    /* Release flb_upstream */
+    if (u != ctx->u) {
+        flb_upstream_destroy(u);
+    }
+
+    return out_ret;
+}
+
+static int compose_payload(struct flb_out_doris *ctx,
+                           const void *in_body, size_t in_size,
+                           void **out_body, size_t *out_size)
+{
+    flb_sds_t encoded;
+
+    *out_body = NULL;
+    *out_size = 0;
+
+    encoded = flb_pack_msgpack_to_json_format(in_body,
+                                              in_size,
+                                              FLB_PACK_JSON_FORMAT_JSON,
+                                              FLB_PACK_JSON_DATE_EPOCH,
+                                              ctx->date_key);
+    if (encoded == NULL) {
+        flb_plg_error(ctx->ins, "failed to convert json");
+        return FLB_ERROR;
+    }
+    *out_body = (void*)encoded;
+    *out_size = flb_sds_len(encoded);
+
+    if (ctx->log_request) {
+        flb_plg_info(ctx->ins, "http body: %s", (char*) *out_body);
+    } else {
+        flb_plg_debug(ctx->ins, "http body: %s", (char*) *out_body);
+    }
+
+    return FLB_OK;
+}
+
+static void cb_doris_flush(struct flb_event_chunk *event_chunk,
+                           struct flb_output_flush *out_flush,
+                           struct flb_input_instance *i_ins,
+                           void *out_context,
+                           struct flb_config *config)
+{
+    int ret = FLB_ERROR;
+    struct flb_out_doris *ctx = out_context;
+    void *out_body;
+    size_t out_size;
+    (void) i_ins;
+
+    char label[256] = {0};
+    int len = 0;
+
+    ret = compose_payload(ctx, event_chunk->data, event_chunk->size,
+                          &out_body, &out_size);
+
+    if (ret != FLB_OK) {
+        if (ret == FLB_ERROR && ctx->log_progress_interval > 0) {
+#ifdef FLB_SYSTEM_WINDOWS
+            InterlockedAdd(&ctx->reporter->failed_rows, event_chunk->total_events);
+#else
+            __sync_fetch_and_add(&ctx->reporter->failed_rows, event_chunk->total_events);
+#endif
```

Comment:
How about creating a function which handles __sync_fetch_add_add like inline function and use it in this plugin. We don't want to spread out `#ifdef` ... `#endif` definitions in Fluent Bit code as much as possible.

Reply:
Thanks for your advice,  I have changed the code.